Feat: ability to validate atoms , `validateAtoms`

barelyhuman opened the pull request:
Adds the `validateAtoms` implementation for overall form atom validation.

Usage example:

```js
import { atomWithValidate, validateAtoms } from "jotai-form";

const nameAtom = atomWithValidate("foo", {
  validate: (v) => v, // individual atom validation (nothing here just for this example)
});

const ageAtom = atomWithValidate(18, {
  validate: (v) => v, // individual atom validation (nothing here just for this example)
});

const formStateAtom = validateAtoms(
  {
    name: nameAtom,
    age: ageAtom,
  },
  async (values) => {
   // validating the entire atom group for a specific form
    await Yup.object()
      .shape({
        name: Yup.string().min(3).required(),
        age: Yup.number().min(18).required(),
      })
      .validate(values);
  }
);

export const Component = () => {
  const [name, setName] = useAtom(nameAtom);
  const [age, setAge] = useAtom(ageAtom);
  const [formState] = useAtom(formStateAtom);

  return (
    <>
      <input
        placeholder="Name"
        value={name.value}
        onChange={(e) => setName(e.target.value)}
      />
      <input
        type="number"
        placeholder="Age"
        value={age.value}
        onChange={(e) => setAge(parseInt(e.target.value,10))}
      />
      <button type="submit" disabled={!formState.isValid}>Submit</button>
    </>
  );
};

```

barelyhuman committed:
base implementation

barelyhuman commented:
@dai-shi  there is a `ts-expect` statement in the code that I'd like you to check. I think it's from the official jotai libraries expectation of the `Getter` type but would need confirmation

barelyhuman committed:
rename and add to entry file

barelyhuman committed:
test: add basic form validation

dai-shi committed:
fix types

dai-shi committed:
simplify types

dai-shi commented:
Types are fixed. You can't have a union type for getter. I'm not sure if it's possible to improve.

dai-shi commented on src/index.ts:
```diff
@@ -1 +1,2 @@
 export { atomWithValidate } from './atomWithValidate';
+export { validateAtom } from './validateForm';
```
```suggestion
export { validateAtom } from './validateAtom';
```
Seems cleaner?

barelyhuman replied:
semantically , there’s going to be more than one atom in most cases?

dai-shi replied:
Can be so. This is not a strong suggestion, so if it's just intentional (rather than a leftover), feel free to leave it as is.

barelyhuman replied:
No, I get concern with the inconsistency it has with everything else but something that doesn't straight point to 1 atom would be a valid replacement.  

cause, if we name it `validateAtom` , instinctively people would just pass it an atom if not working in an IDE with language servers. 

`validateForm` is a placeholder but, also seems like there isn't a simple name for something that has a ValueGetter inside as it's different than most jotai API's

dai-shi replied:
I wouldn't usually export types and probably have it in a separate file if types are used in two files.

Actually, I don't mind much about the internal file names, which can be changed at any time. But, maybe we would revisit the API name `validateAtom`, because ValueGetter is a new one as you say.

barelyhuman replied:
What if, we change the API to be like so 

```js 
const fNameAtom = atomWithValidate("", { validate: (v) => v });
const lNameAtom = atomWithValidate("", { validate: (v) => v });

validateAtomFamily(
  { firstName: fNameAtom, lastName: lNameAtom },
  { validate: (v) => v }
);
// or
validateAtomGroup(
  { firstName: fNameAtom, lastName: lNameAtom },
  { validate: (v) => v }
);


```

dai-shi replied:
"atom group" sounds better ("atom family" is too confusing with `atomFamily`).

```js
validateAtoms(async (get) => {
  const values = getAtomGroup(get)({ firstName: fNameAtom, lastName: lNameAtom });
  await validate(values);
});
```
can be my preference, but it might not be developer friendly.

Wait, we have `waitForAll` util.
```js
validateAtoms(
  { firstName: fNameAtom, lastName: lNameAtom }, // or [fNameAtom, lNameAtom]
  async (values) => ...
)
```
would match with it. Then, returning the combined atom values makes more sense. So, it's not just "validate"...

Hmm, let's think about this more.
```js
const fNameAtom = atomWithValidate("", { validate: (v) => v });
const lNameAtom = atomWithValidate("", { validate: (v) => v });

const combinedAtom = combineAtoms(
  { firstName: fNameAtom, lastName: lNameAtom },
  { validate: (v) => v }
);
```
It's almost like your proposition. Not sure if "combine" is the right word.

dai-shi replied:
On second thought, `validateAtoms` would sound more consistent with `atomWithValidate`.
```js
const personNameAtom = validateAtoms(
  { firstName: fNameAtom, lastName: lNameAtom },
  { validate: async (values) => ... }
);
```

barelyhuman replied:
I spend doing the same thing on paper 
and I think `validateAtomGroup` or `validateAtoms` and the API will be like the last comment 
```js
const personNameAtom = validateAtoms(
  { firstName: fNameAtom, lastName: lNameAtom },
  { validate: async (values) => ... }
);
```

dai-shi replied:
Sounds great. Now we are almost back to one of original propositions at some point. Would you update the implementation?

barelyhuman replied:
Yeah, on it

barelyhuman committed:
refactor: validateAtoms API

will change the API to avoid adding a helper to reduce developer work
for basic interaction

dai-shi commented on src/validateAtoms.ts:
```diff
@@ -0,0 +1,72 @@
+
+import type { CommonState } from './atomWithValidate';
+
+export type GetValues = <Value>(labeledAtoms: LabeledAtoms<Value>) => {
+  [k: string]: Value;
+};
```
No longer used, right?

barelyhuman replied:
good catch

dai-shi commented on src/validateAtoms.ts:
```diff
@@ -0,0 +1,72 @@
+export type GetValues = <Value>(labeledAtoms: LabeledAtoms<Value>) => {
+  [k: string]: Value;
+};
+export type Validator = <Value>(values: { [k: string]: Value }) => Promise<any>;
```
Maybe something like this?
```suggestion
export type Validator = <Values extends Record<string, unknown>>(values: Values) => Promise<any>;
```

dai-shi commented on src/validateAtoms.ts:
```diff
@@ -0,0 +1,72 @@
+};
+export type Validator = <Value>(values: { [k: string]: Value }) => Promise<any>;
+
+export type ValidatorState = {
```
We want to return "values" too, no? I'm taking down one of my past ideas.

barelyhuman replied:
We can return them, would make it consistent with the `atomWithValidate` API 

though, there's no actual use-case of the values here since we don't manage an atom with them.

dai-shi replied:
Like you said for a validator library, combined values can be useful for a certain UI library.
`waitForAll` is similar, in some cases we don't use the return value and in other cases we use it.

barelyhuman replied:
I've made the change, 
just in case we extended on this with something else later

barelyhuman replied:
Just have to write docs and examples of usage now

dai-shi replied:
> I've made the change

Not pushed yet??

barelyhuman replied:
it’s in the same commit 
https://github.com/jotai-labs/jotai-form/pull/5/commits/2c02a50da9bb336474eee696725f768524247300

dai-shi replied:
okay, that's different from my expectation. I want it to be closer to waitForAll mental model (of mine...)

barelyhuman replied:
oh, i’ll refer to the source of that and see what you mean

dai-shi reviewed (commented):
quick review

barelyhuman committed:
fix: remove types, consistent validator function

remove the un-needed types from the API change
and also make the validator function consistent
with atomWithValidate

dai-shi commented on src/validateAtoms.ts:
```diff
@@ -0,0 +1,70 @@
+
+export type Validator = <Values extends Record<string, unknown>>(
+  values: Values,
+) => Promise<Values>;
```
```suggestion
) => Promise<unknown>;
```


and,

```tsx
type State<Values extends Record<string, unknown>> = {
  values: Values;
} & ValidatorState
```
(In the end, I'd combine it into one type)

dai-shi commented on src/validateAtoms.ts:
```diff
@@ -0,0 +1,70 @@
+      }),
+    );
+
+    return validator(values);
```
```suggestion
    await validator(values);
    return values;
```

Oh, okay, this alone doesn't work because of loadable.
We'd need one more internal atom.

barelyhuman replied:
So, if I understand correctly 
the `validation` output when used as 
```js
// validation  would have 
// validation.isValid,validation.isValidating,validation.values, validation.error ?  
const [validation] = useAtom(validationAtom)
```

dai-shi replied:
yes

dai-shi reviewed (commented):
Not sure the explanation is enough, but let it go.

barelyhuman committed:
add values into the atom

barelyhuman commented:
Do not merge even if it suffices, have to write tests  `async` validation

dai-shi commented on src/validateAtoms.ts:
```diff
@@ -0,0 +1,88 @@
+
+export type Validator = <Values extends Record<string, unknown>>(
+  values: Values,
+) => Promise<unknown>;
```
Maybe this should be `unknown | Promise<unknown>` or `void | Promise<void>`. We support sync validator, right?

barelyhuman replied:
that totally depends on the `loadable` API but I think the sync one should be working, i'll check in the evening

barelyhuman replied:
done.
Though, there's an issue with using mixed typed atoms right now. The types for validator and labelled atoms needs to be fixed for that. 

On it next.

dai-shi reviewed (approved):
other than that, the code looks okay.

barelyhuman committed:
chore: type simplification and test

will add sync validator function in types
and adds a test trying it out with an async test library

barelyhuman committed:
fix: infer generic types

will infer the atom's internal generic
and use that to generate the type
sent to the validator

dai-shi commented on src/validateAtoms.ts:
```diff
@@ -0,0 +1,89 @@
+  SetStateAction<Value>
+>;
+
+type State<Values extends Record<string, unknown>> = {
```
Should `unknown` be `any` like L28? I always wonder which is better, and how it's different.

barelyhuman replied:
`any` would allow the usage be a little more free, for example with `any` the variable `a` could do the following 
```ts
a.toUpperCase() // even when a is not a string
```

on the other hand `unknown` would raise an error 
```ts
    a.toUpperCase()
/// ^ `a` is unknown or property `toUpperCase` doesn't exist on unknown
```

and nice point, this should actually inherit the value types the same way Validator is

barelyhuman replied:
It's inferring the types already, 
anything should be fine on this type, i'd leave it to be `unknown` so people can add type checks 
L28 uses `any` because we need to infer the types from the passed atom group, adding `unknown` there would mean the dev would have to manually typecast each atom, which wouldn't be nice. 

as for using `unknown` here would force them to check types in the Validator which is needed if they are writing validations. Though, in most cases jotai is used with primitive types (number, strings,etc) the inference should already help with this.

dai-shi replied:
thanks for the explanation!

dai-shi commented on src/validateAtoms.ts:
```diff
@@ -0,0 +1,89 @@
+  labeledAtoms: AtomGroup,
+  validator: Validator<Keys, inferGeneric<Vals>>,
+) => {
+  const $getSourceAtomVals = (get: Getter) => {
+    const values = Object.fromEntries(
+      Object.entries(labeledAtoms).map(([k, v]) => {
+        const atomValue = get(v);
+        return [k, atomValue.value];
+      }),
+    );
+    return values as Record<Keys, inferGeneric<Vals>>;
+  };
```
I thinks it's better to make this an atom.
```suggestion
  const valsAtom = atom((get) => {
    const values = Object.fromEntries(
      Object.entries(labeledAtoms).map(([k, v]) => {
        const atomValue = get(v);
        return [k, atomValue.value];
      }),
    );
    return values as Record<Keys, inferGeneric<Vals>>;
  });
```

barelyhuman replied:
too many listeners on the store? 
cause it's being used in 2 places as a simple helper right now 

```ts
const valsAtom = atom((get: Getter) => {
    const values = Object.fromEntries(
      Object.entries(labeledAtoms).map(([k, v]) => {
        const atomValue = get(v);
        return [k, atomValue.value];
      }),
    );
    return values as Record<Keys, inferGeneric<Vals>>;
  });

  const baseAtom = atom(async (get) => {
    // extract value from each atom and assign to the given key as label
    return validator(get(valsAtom));
  });

  const normalizerAtom = atom((get) => {
    const values = get(valsAtom);
    const state = get(loadable(baseAtom));
    return {
      ...state,
      values,
    };
  });

  const derv = atom((get) => {
    const validatorState = get(normalizerAtom);
//.... other code
})
```

dai-shi replied:
`valsAtom` can be more efficient because it caches the result, than a simple helper.

barelyhuman replied:
makes sense, pushed

dai-shi commented on src/validateAtoms.ts:
```diff
@@ -0,0 +1,89 @@
+} & ValidatorState;
+
+export const validateAtoms = <
+  AtomGroup extends Record<string, AtomWithValidation<any>>,
+  Keys extends keyof AtomGroup,
+  Vals extends AtomGroup[Keys],
```
Could this be the other way around?
```suggestion
  Vals extends Record<string, unknown>,
  AtomGroup extends { [key in keyof Vals]: AtomWithValidation<Vals[key]>>,
```

and in the code:
```ts
type Keys = keyof Vals;
```

barelyhuman replied:
I think I tried this and TS wasn't inferring the passed object's types, I'll try it again with the above suggestion and ping back

dai-shi replied:
I think it can actually be something like this:
```ts
const validateAtoms = <Vals>(
  labeledAtoms: { [key in keyof Vals]: AtomWithValidation<Vals[key]>>,
  validator: Validator<Vals>,
) => {
```
hm, maybe you are right, it can't infer types from `labeledAtoms`.

barelyhuman replied:
Yeah,
Both won't work, the type inheritance needs it to point on the passed object to infer it as a generic

dai-shi replied:
makes sense.

dai-shi reviewed (commented):
final review.

barelyhuman committed:
refac: value getter to atom

dai-shi commented on src/validateAtoms.ts:
```diff
@@ -0,0 +1,88 @@
+  });
+
+  const derv = atom((get) => {
+    const validatorState = get(normalizerAtom);
```
```suggestion
    const values = get(valsAtom);
    const state = get(loadable(baseAtom));
    const validatorState = {
      ...state,
      values,
    };
```
Does this work?

barelyhuman replied:
the test's say it's working, 
that's nice, 1 listener reduced.

barelyhuman replied:
pushed with other simplifications

barelyhuman committed:
simplify

- remove un-needed variables
- use switch instead to avoid case checks

dai-shi reviewed (approved):
LGTM

Please update the PR description.

Feel free to merge it and ship it.

BenBeattieHood commented:
Oh wow! I go away for a break, I come back and you've solved this 🤩 Love your work, thank you 👍
